api/server.go

```diff
@@ -0,0 +1,293 @@
+// ie. "X-XSS-Protection" - "1; mode=block"
+// ie. "Server" - "lbry.io"
+// ie. "Referrer-Policy" - "same-origin"
+// ie. "Strict-Transport-Security" - "max-age=31536000; preload"
+// ie. "Access-Control-Allow-Origin" -"<header.Origin>"
+// ie. "Access-Control-Allow-Methods" - "GET, POST, OPTIONS"
+var HeaderSettings map[string]string
+
+// LogError Allows specific error logging for the server at specific points.
+var LogError = func(err error) {}
+// LogInfo Allows for specific logging information.
+var LogInfo  = func(info string) {}
+
+// TraceEnabled Attaches a trace field to the JSON response when enabled.
+var TraceEnabled = false
+
+var errAuthenticationRequired = errors.Base("authentication required")
+var errNotAuthenticated = errors.Base("could not authenticate user")
+var errForbidden = errors.Base("you are not authorized to perform this action")
+
+// StatusError represents an error with an associated HTTP status code.
+type StatusError struct {
+	Status int
+	Err    error
+}
+
+// Allows StatusError to satisfy the error interface.
+func (se StatusError) Error() string {
+	return se.Err.Error()
+}
+
+// Response is returned by API handlers
+type Response struct {
+	Status      int
+	Data        interface{}
+	RedirectURL string
+	Error       error
+}
+
+// Handler handles API requests
+type Handler func(r *http.Request) Response
+
+func (h Handler) callHandlerSafely(r *http.Request) (rsp Response) {
+	defer func() {
+		if r := recover(); r != nil {
+			err, ok := r.(error)
+			if !ok {
+				err = errors.Err("%v", r)
+			}
+			rsp = Response{Error: errors.Wrap(err, 2)}
+		}
+	}()
+
+	return h(r)
+}
+
+func (h Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	// Set header settings
+	if HeaderSettings != nil {
+		//Multiple readers, no writers is okay
+		for key, value := range HeaderSettings {
+			w.Header().Set(key, value)
+		}
+	}
+
+	// Stop here if its a preflighted OPTIONS request
+	if r.Method == "OPTIONS" {
+		return
+	}
+
+	rsp := h.callHandlerSafely(r)
+
+	if rsp.Status == 0 {
+		if rsp.Error != nil {
+			if statusError, ok := rsp.Error.(StatusError); ok {
+				rsp.Status = statusError.Status
+			} else if errors.Is(rsp.Error, errAuthenticationRequired) {
+				rsp.Status = http.StatusUnauthorized
+			} else if errors.Is(rsp.Error, errNotAuthenticated) || errors.Is(rsp.Error, errForbidden) {
+				rsp.Status = http.StatusForbidden
+			} else {
+				rsp.Status = http.StatusInternalServerError
+			}
+		} else if rsp.RedirectURL != "" {
+			rsp.Status = http.StatusFound
+		} else {
+			rsp.Status = http.StatusOK
+		}
+	}
+
+	success := rsp.Status < http.StatusBadRequest
+
+	consoleText := r.RemoteAddr + " [" + strconv.Itoa(rsp.Status) + "]: " + r.Method + " " + r.URL.Path
+	// query := r.URL.Query()
+	// if len(query) > 0 {
+	// 	queryJSON, _ := json.Marshal(query)
+	// 	consoleText += " " + string(queryJSON)
+	// }
+	if success {
+		LogInfo(color.GreenString(consoleText))
```

Comment:
dont use `color` here. that's up to the LogInfo / LogError implementation.

In fact, LogInfo should probably just be passed the request and response, so it can do whatever it wants. And LogError should be passed request, response, and error. There's no need for us to build this string here. What if another implementation wants different data.

Reply:
good point.